### PR TITLE
PLAT-165: Change CoreGroup model

### PR DESCRIPTION
**Pull request by docktorrr**
## Purpose
We need to change CoreGroup (main model responsible for permissions management).
Later we need to replace Group and WorkflowTeam permissions management with this one.
And some refactoring was made together with that.

## Approach
1. Remove Role model
2. Remove user_permissions overriding in CoreUser (after switching to custom user model we don't need this overriding).
3. Change a bit an order of models in workflow/models.py to make dependant models defined after its dependancies.
4. Main part: Change CoreGroup model (see the docstring). Now it has no organization field, the organization is taken from workflowlevel1/2.
5. Change CoreGroup serializer, viewset and tests according to the model changes.
6. Make some IsOrgMember permission refactoring to make it more universal.

### CODE REVIEW

File: workflow/migrations/0002_auto_20190328_1231.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        ),
+        migrations.RemoveField(
+            model_name='coregroup',
+            name='permissions',
+        ),
+        migrations.AddField(
+            model_name='coregroup',
+            name='permissions',
+            field=models.PositiveSmallIntegerField(default=15, help_text='Decimal integer from 0 to 15 converted from 4-bit binary, each bit indicates permissions for CRUD', verbose_name='Permissions'),
+        ),
```
Comment (jefmoura) [inline workflow/migrations/0002_auto_20190328_1231.py]: I like the approach, we just need to describe all the possibilities in the docs.

Comment (jefmoura) [reply workflow/migrations/0002_auto_20190328_1231.py]: Shouldn't the default be 4? (Read-only)

File: workflow/models.py [hunk 2]
```diff
@@ -244,6 +201,100 @@ def __str__(self):
...
+    def organization(self) -> Union[Organization, None]:
+        if self.workflowlevel2:
+            return self.workflowlevel2.organization
+        if self.workflowlevel1:
+            return self.workflowlevel1.organization
+        return None
+
+    @property
+    def display_permissions(self) -> str:
+        return '{0:b}'.format(self.permissions if self.permissions < 16 else 15)
```
Comment (jefmoura) [inline workflow/models.py]: I think it'd be nice and clearer if we add  a length to it, like `{0:04b}`, so we always see the 4 bits

File: workflow/permissions.py [hunk 3]
```diff
@@ -122,17 +121,15 @@ def has_object_permission(self, request, view, obj):
...
-            elif obj.__class__ in [Organization]:
+            if obj.__class__ in [Organization]:
                 return obj.id == user_org
             elif obj.__class__ in [WorkflowTeam]:
                 if request.user.is_org_admin():
-                    return obj.__class__.objects.filter(workflow_user__organization=user_org, id=obj.id).exists()
+                    return obj.workflow_user and obj.workflow_user.organization == user_org  # ?
                 else:
-                    return obj.__class__.objects.filter(workflowlevel1__organization=user_org, id=obj.id).exists()
+                    return obj.organization.id == user_org
```
Comment (jefmoura) [inline workflow/permissions.py:130]: WorkflowTeam doesn't have an organization, so this check will raise an error.

Comment (docktorrr) [reply workflow/permissions.py:130]: We have it now as a model's property: https://github.com/Humanitec/bifrost/blob/refactor-coregroup/workflow/models.py#L333

File: workflow/models.py (hunk 2 above)
Comment (docktorrr) [reply workflow/models.py]: Fixed

File: workflow/migrations/0002_auto_20190328_1231.py (hunk 1 above)
Comment (docktorrr) [reply workflow/migrations/0002_auto_20190328_1231.py]: Changed to 4

**Comment by docktorrr** [comment]
@jefmoura Changed relations between CoreGroup and WorkflowLevel1/2 to m2m.

**Review by jefmoura — APPROVED** [review]

**Pull request merged**